### PR TITLE
Refactor auth_credential table to STI

### PR DESCRIPTION
**Pull request by brendanshean**
## WHAT
Reorganize the auth_credential table to use STI with subclasses based on the provider names.

## WHY
With the addition of another provider type (i.e. Canvas) it would simplify logic to move the respective providers into their own subclasses.

## HOW
1. Create subclasses `CanvasAuthCredential`, `CleverDistrictAuthCredential`, `CleverLibraryAuthCredential`, `GoogleAuthCredential`
1. Update the places where these AuthCredentials are saved to include these subtypes.
1. Add a backfill script to be run after migration.   Once the backfill is complete, a subsequent PR can add a `null: false` to the type column.

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | (N/A or Yes)

### CODE REVIEW

**Review by anathomical — APPROVED** [review]
I totally dig this refactor.  Should make keeping some of this stuff much easier to maintain and expand moving forward.

File: services/QuillLMS/app/models/auth_credential.rb [hunk 1]
```diff
@@ -26,46 +27,27 @@
...
-    google_provider? && !refresh_token_valid?
-  end
-
-  def google_authorized?
-    google_provider? && refresh_token_valid?
-  end
-
-  def google_provider?
-    provider == GOOGLE_PROVIDER
+  def canvas_authorized?
```
Comment (dandrabik) [inline services/QuillLMS/app/models/auth_credential.rb:38]: If we are going the STI approach, we might think about converting these methods to more generic method like `authorized?` and changing some of the logic about how these are called. That might be better as a separate possible refactor, given the diff size.

File: services/QuillLMS/app/models/auth_credential.rb [hunk 2]
```diff
@@ -26,46 +27,27 @@
 #
 class AuthCredential < ApplicationRecord
   belongs_to :user
```
Comment (dandrabik) [inline services/QuillLMS/app/models/auth_credential.rb:27]: Also, for the STI route, I think we'd want to change the relationship on `User` to be `has_one :google_auth_credential` (one for each). That could be better as a separate refactor, but then you could do `user.google_auth_credential&.authorized?` and/or move the authorization methods to `user` where they might read more naturally, e.g.
```ruby
# In User
def google_authorized?
  google_auth_credential&.authorized?
end

File: services/QuillLMS/app/services/canvas_integration/auth_credential_saver.rb [hunk 3]
```diff
@@ -29,10 +27,10 @@ def run
 
     private def auth_credential
       @auth_credential ||=
-        AuthCredential.create!(
+        CanvasAuthCredential.create!(
```
Comment (dandrabik) [inline services/QuillLMS/app/services/canvas_integration/auth_credential_saver.rb:30]: I don't have strong feelings either way, but one thing to consider (take or leave): Since you now have a subclass that you can put Canvas-specific logic in, do you still need this Saver service? Or would this be better moving this to the Rails model with validations?.

File: services/QuillLMS/db/migrate/20230622125712_add_type_to_auth_credential.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddTypeToAuthCredential < ActiveRecord::Migration[6.1]
+  def change
```
Comment (dandrabik) [inline services/QuillLMS/db/migrate/20230622125712_add_type_to_auth_credential.rb]: You might consider putting this migration and the data migration in a separate PR and deploying that first to cut down on potential errors and/or downtime (All of this logic depends on type being there and populated). Then you can deploy the code and re-run the data migration for the last few records.

**Review by dandrabik — APPROVED** [review]
Overall, looks good. I left a few comments "for your consideration" but nothing blocking. I do think it would be better to deploy the migration first, since this code needs that column right away.

Nice work on this!

File: services/QuillLMS/app/services/google_integration/user.rb [hunk 5]
```diff
@@ -43,10 +43,11 @@ def update_or_initialize
 
   private def auth_credential_attributes(user)
     {
-      provider:      'google',
-      refresh_token: profile.refresh_token || user&.auth_credential&.refresh_token,
-      expires_at:    profile.expires_at,
       access_token:  profile.access_token,
+      expires_at:    profile.expires_at,
+      refresh_token: profile.refresh_token || user&.auth_credential&.refresh_token,
```
Comment (happythenewsad) [inline services/QuillLMS/app/services/google_integration/user.rb]: This is orthogonal to this PR, but can we get rid of a code path and just do 

`refresh_token: profile.refresh_token` ?

**Review by happythenewsad — APPROVED** [review]

File: services/QuillLMS/app/services/google_integration/user.rb (hunk 5 above)
Comment (happythenewsad) [reply services/QuillLMS/app/services/google_integration/user.rb]: Also, this line will need to be updated if you implement Dan's suggestion of adding `has_one` associations in `User.rb` for each STI class

File: services/QuillLMS/db/migrate/20230622125712_add_type_to_auth_credential.rb (hunk 4 above)
Comment (brendanshean) [reply services/QuillLMS/db/migrate/20230622125712_add_type_to_auth_credential.rb]: That's a good point.  I will pull it out in a separate PR.

File: services/QuillLMS/app/services/canvas_integration/auth_credential_saver.rb (hunk 3 above)
Comment (brendanshean) [reply services/QuillLMS/app/services/canvas_integration/auth_credential_saver.rb:30]: I like the idea, but I don't want the scope of this to grow too much.  I have an overall larger refactor for later that I'm imagining with Clever/Canvas/Google so I will add this as a note to that.

File: services/QuillLMS/app/models/auth_credential.rb (hunk 2 above)
Comment (brendanshean) [reply services/QuillLMS/app/models/auth_credential.rb:29]: I was thinking about the `has_one :canvas_auth_credential`, `has_one :google_auth_credential`, etc. but felt that would actually change how we use auth_credentials sitewide since it would allow for the possibility of multiple auth_credentials.  We should discuss offline

Comment (brendanshean) [reply services/QuillLMS/app/models/auth_credential.rb:27]: I was actually thinking about the `has_one :canvas_auth_credential`, `has_one :google_auth_credential`, etc. yesterday, but felt that would actually change how we use auth_credentials sitewide since it would allow for the possibility of multiple auth_credentials.  We should probably discuss offline

File: services/QuillLMS/app/services/google_integration/user.rb (hunk 5 above)
Comment (brendanshean) [reply services/QuillLMS/app/services/google_integration/user.rb]: Good call.  I will remove that.

File: services/QuillLMS/app/models/auth_credential.rb (hunk 1 above)
Comment (brendanshean) [reply services/QuillLMS/app/models/auth_credential.rb:38]: I was thinking about a generic `authorized?` but hadn't gotten into specifics.  I like your solution.

File: services/QuillLMS/app/models/auth_credential.rb (hunk 2 above)
Comment (brendanshean) [reply services/QuillLMS/app/models/auth_credential.rb:27]: After discussing offline, Dan and I agreed that this is a good potential solution long term, but out of scope for this project.
